lms/djangoapps/courseware/tabs.py

```diff
@@ -383,6 +383,14 @@ def get_course_tab_list(user, course):
     # further investigation since CourseTabList.iterate_displayable returns
     # Static Tabs that are not returned by the CourseTabPluginManager.
     course_tab_list.sort(key=lambda tab: tab.priority or float('inf'))
+
+    if course_tab_list:
```

Comment:
Why `if course_tab_list:`?

More broadly, why this method? Feels like a very surprising side effect. Was it just a method you found that both courseware and course home called? 

I might just have two calls (one for each metadata), rather than having such a surprising side effect for this method. And that would go down to one call anyway when https://openedx.atlassian.net/browse/AA-1018 is fixed.

Reply:
doesn't really make sense to record activity if a user can't see any tabs

yes but more importantly this also covers the tabs that are not in the MFE and the courseware that is not in the MFE, since we need to cover those things for this as well as the MFE

Reply:
> doesn't really make sense to record activity if a user can't see any tabs

When would there not be any tabs?

> yes but more importantly this also covers the tabs that are not in the MFE and the courseware that is not in the MFE, since we need to cover those things for this as well as the MFE

Interesting. I'm still skeptical that this is a reasonable side effect for this method. I would never in a million years guess that `get_course_tab_list` also wrote to the database that the user did something that day. Which suggests that it's in the wrong place to me.

If you do keep it here, I think you need to add blaring sirens to the docstring about this. But I would have a think about if there isn't a more natural place to put it.

Reply:
if course_tab_list is empty there would be no tabs - presumably not an expected case, but seems technically possible.

i do agree that is confusing, though i think hooking into the backend code for a get request would always be an unexpected side effect

i guess i could do the following:
rather than having one location, i could
1. Create a call alongside the call to this one in course_navigation to cover the non-mfe experience
2. Add a call in the courseware_api
3. Add a call in the course home api
Not sure if that solves for your concern about modification of the database being an unexpected side effect?
(though we do do that as well for other things such as the streak)

Reply:
I've updated the code to implement what i described in the above comment

Reply:
> if course_tab_list is empty there would be no tabs - presumably not an expected case, but seems technically possible.

OK... But that never happens right? And even if it did, the user still visited us and we should record that, yeah?

> i think hooking into the backend code for a get request would always be an unexpected side effect

I disagree with that. And if you think that, do you view this whole feature as a bad design?

If we're talking a RESTful CRUD style API, then yes - modifying the object under discussion on a GET would certainly be unexpected. But I would still expect a server log file entry to be generated somewhere for that GET. Maybe other analytics recorded too. And this feature is basically analytics.

If we take a step up to a less-CRUD like API (like this) or an html page browser GET, I would absolutely not be surprised at side effects. Maybe creating stuff lazily as the user views pages. Or much more common, lots and lots of analytic records get generated on each view. New Relic, Segment, django, caching, etc.

But it's when we get to the level of a Python method, styled as a code-level GET like `get_course_tab_list` - that's where side effect database writes/analytics become more surprising to me.

> rather than having one location, i could

That approach feels less surprising to me (having an endpoint view directly record the activity).

I'm also open to some other method that those paths all happen to call, if it had a name that could imply what we are doing here. Like, if they happened to all call some method `do_analytics` which sent a segment event and now this table, and who knows what else. You could make such a method if it doesn't exist, with just this one call in it for now, but that seems maybe too YAGNI.

Reply:
(didn't see you had already changed it by the time I made my comment)